SoftClip.h

```diff
@@ -0,0 +1,50 @@
+#ifndef SOFT_CLIP_H
+#define SOFT_CLIP_H
+
+#include "AudioOutput.h"
+
+template<unsigned int SMOOTHING, typename T = AudioOutputStorage_t>
+class SoftClip
+{
+  private:
+    const unsigned int max_threshold = (unsigned int) SMOOTHING * PI / 2;
+    T lookup_table[(unsigned int) (SMOOTHING * PI / 2)];
+    T max_value;
+
+  public:
+    SoftClip(T max_val)
+    {
+      max_value = max_val;
+      for (unsigned int i = 0; i < max_threshold; i++)
+      {
+        lookup_table[i] = SMOOTHING * sin(1.*i / SMOOTHING);
```

Comment:
Note that the calculation is done only once, probably before audio starts if the user does not try to have "local" version of this class in the main loop, hence it is very fast. Memory is more my concern however, as you pointed, the main problem of the tables is that they are 8bits only whereas this can compute a soft clip on *any* resolution, which is crucial for the final flattening (when reaching saturation) otherwise there are huge jumps in the output values which are more like an hard clip. We could potentially interpolate them though… As you probably guessed now, I am usually optimizing computing time at the expense of memory.